### PR TITLE
Account for Collection Prefs when cloning/carry forwarding

### LINKED ISSUE #3278: Cloning incorrectly duplicates preparations
**Steps to Recreate:**
1. Go to the following URL and log in as `spvertpaleoadmin`:
    https://kuvertpaleo11123-edge.test.specifysystems.org/specify/view/collectionobject/2/
2. See that Preparations has only one item
3. Click **Clone**
4. See that there are now 3 preparations
5. Click **Save**
6. See error

`(1048, "Column 'PrepTypeID' cannot be null")`

**Before clone:**
image

**After clone:**
https://user-images.githubusercontent.com/37256050/228937966-fb64d82b-25a6-43bb-8534-0e89ee88bbf1.mov

```
IntegrityError at /api/specify/collectionobject/ (1048, "Column 'PrepTypeID' cannot be null") Request Method: POST Request URL: http://kuvertpaleo11123-edge.test.specifysystems.org/api/specify/collectionobject/ Django Version: 3.2.15 Python Executable: /opt/specify7/ve/bin/python3.8 Python Version: 3.8.0 Python Path: ['/opt/specify7', '/opt/specify7', '/opt/specify7/ve/bin', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '/opt/specify7/ve/lib/python3.8/site-packages', '/opt/specify7'] Server time: Thu, 30 Mar 2023 19:00:51 -0500 Installed Applications: ('django.contrib.sessions', 'django.contrib.staticfiles', 'django.contrib.contenttypes', 'django.contrib.auth', 'specifyweb.specify', 'specifyweb.permissions', 'specifyweb.accounts', 'specifyweb.stored_queries',
```

[Specify 7 Crash Report - 2023-03-30T19_01_22.353Z.txt](https://github.com/specify/specify7/files/11115130/Specify.7.Crash.Report.-.2023-03-30T19_01_22.353Z.txt)

`v7.8.8`

**Reported By:** KU Vert Paleo and @acbentley

### PR DESCRIPTION
**Pull request by melton-jason**
Fixes #3278 

TO TEST:
- Ensure the `CO_CREATE_PREP_<COLLECTION>` preference is defined in Remote Preferences
  - Replace `<COLLECTION>` with the id of the collection you are testing this in
- Clone and Carry Forward a CollectionObject which has existing preparations. 
- Clone and Carry Forward a CollectionObject which does not have preparations

- Test the other "collection preferences" related to automatic resource creation/initialization which can be defined in Remote Preferences:
  - CO_CREATE_COA_<COLLECTION> (Collection Object Attribute)
  - CO_CREATE_DET_<COLLECTION> (Determination)

### CODE REVIEW

**Review by CarolineDenis — APPROVED** [review]

File: specifyweb/frontend/js_src/lib/components/DataModel/__tests__/domain.test.ts [hunk 1]
```diff
@@ -98,3 +99,66 @@ describe('fetchCollectionsForResource', () => {
...
+     * When cloning the resource, an empty CollectionObjectAttribute is created as well, causing the 'expected inline data for dependent field' warning from /DataModel/resourceApi.js
+     */
+    jest.spyOn(console, 'warn').mockImplementation();
+    const cloned = await collectionObject.clone(true);
+
+    await expect(
+      cloned
+        .rgetCollection('preparations')
+        .then((collection) => collection.models.length)
+    ).resolves.toBe(1);
```
[the remark this answers is not in the document]
Comment (melton-jason) [reply specifyweb/frontend/js_src/lib/components/DataModel/__tests__/domain.test.ts:162]: Can you further explain what you mean by this? 
When the preference (for this example, I will use the CO_CREATE_PREP) is enabled for the currently logged in collection, there are two cases to consider when initializing a CollectionObject. 

Either: 
- The CollectionObject does not have any preparations
  - In which case create a new one
- The CO  already has preparations
  - In which case there should be no new preparations 

Both of these cases are handled by these tests. 
Do you mean test when the preference is not configured?

[the remark this answers is not in the document]
Comment (realVinayak) [reply specifyweb/frontend/js_src/lib/components/DataModel/__tests__/domain.test.ts:162]: > Both of these cases are handled by these tests.
They aren't? Doesn't this test check if cloned collection object has preparations? Maybe I don't have have context. Can you quote the line where you are checking that cloned collection object does not have preparations?

**Review by realVinayak — CHANGES_REQUESTED** [review]

File: specifyweb/frontend/js_src/lib/components/DataModel/__tests__/domain.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (melton-jason) [reply specifyweb/frontend/js_src/lib/components/DataModel/__tests__/domain.test.ts:162]: > - The CollectionObject does not have any preparations
>   - In which case create a new one

https://github.com/specify/specify7/blob/a09ecc062ed1f6f7513a6bb194974214bd54ee37/specifyweb/frontend/js_src/lib/components/DataModel/__tests__/domain.test.ts#L121-L126

- - - 

> - The CO  already has preparations
>   - In which case there should be no new preparations 

We add a preparation to the CollectionObject here: 

https://github.com/specify/specify7/blob/a09ecc062ed1f6f7513a6bb194974214bd54ee37/specifyweb/frontend/js_src/lib/components/DataModel/__tests__/domain.test.ts#L141-L150

After cloning, the CollectionObject still only has one preparation:

https://github.com/specify/specify7/blob/a09ecc062ed1f6f7513a6bb194974214bd54ee37/specifyweb/frontend/js_src/lib/components/DataModel/__tests__/domain.test.ts#L158-L162

Previously, for this case the cloned CollectionObject would have two or three preparations; the preparation that had existed before, as well as an empty preparation which was created when the resource was initialized (hence the issue). 

> Can you quote the line where you are checking that cloned collection object does not have preparations?
In this case, the cloned CollectionObject _should_ have preparations: the preparation which existed on the original, pre-cloned CollectionObject. 

If the preference is set, there is no case in which a cloned CollectionObject should have no preparations, right?

**Review by pashiav — APPROVED** [review]
I tested for both the Clone and Carry Forward features and they both do what is expected of them! No incorrect duplicates. Adding new determinations also works. Looks good!
**Collection object with No Preparations:** 
issue-3278-no prep
**Collection object with Existing Preparations:**
issue-3278-prep

Added lines in remote pref:
Screen Shot 2023-11-09 at 10 23 08 AM

**Comment by specifysoftware** [comment]
This pull request has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/specify-7-9-2-release-announcement/1452/1

**Pull request merged**